staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go

```diff
@@ -38,17 +44,65 @@ var (
 		},
 		[]string{"name", "reason"},
 	)
-	unavailableGauge = metrics.NewGaugeVec(
-		&metrics.GaugeOpts{
-			Name:           "aggregator_unavailable_apiservice",
-			Help:           "Gauge of APIServices which are marked as unavailable broken down by APIService name.",
-			StabilityLevel: metrics.ALPHA,
-		},
+
+	unavailableGaugeDesc = metrics.NewDesc(
+		"aggregator_unavailable_apiservice",
+		"Gauge of APIServices which are marked as unavailable broken down by APIService name.",
 		[]string{"name"},
+		nil,
+		metrics.ALPHA,
+		"",
 	)
 )
+var registerMetrics sync.Once
+
+// Register registers metrics for status controller.
+func Register(apiServiceLister listers.APIServiceLister) {
+	registerMetrics.Do(func() {
+		legacyregistry.CustomMustRegister(newAPIServiceStatusCollector(apiServiceLister))
```

Comment:
while touching this, can we turn legacyregstry into a parameter of Register? Now it's half a global construct and the other half isn't.

Reply:
the `sync.Once` does not play well with this :-/ Maybe the `sync.Once` belongs where we call Register? These globals are ugly....

Reply:
legacyregistry is a package here and the global registry it uses isn't exported.

Reply:
I would agree with you that this isn't very clean, but I mostly followed what's done in other packages.

I don't think it is meaningful to move the `sync.Once`  to where we call Register. It is used to prevent registering the same metrics multiple times into the global registry. If such a thing occurs, it would result in a panic :/. To me, it should be very local to this Register function and thus not be moved elsewhere.
Also, we won't be able to turn it into a local variable as it would make it possible to register the metrics multiple times.

The correct approach here would be to get rid of the global registry, but this would require a bigger rewrite.